### PR TITLE
Cleanup examples

### PR DESCRIPTION
**Pull request by xnuohz**
- Combine 15 examples related to cora into 1 file
- Update related docs, github workflow, readme

### CODE REVIEW

**Review by puririshi98 — CHANGES_REQUESTED** [review]
code changes look good at first review, but for such an extensive change i would ask that you run every one of the deleted examples, and then the new version with its corresponding args for this planetoid example. I would ask that you share those logs and also check yourself to make sure that the new version is as good or better in terms of accuracy, speed, and readability. and if not for any reason try to understand why. This is a very valuable PR but we want to be careful with something like this. 

If compute is an issue, i would ask that you write a bash script for me that will automatically run the above experiments and have the logs saved to some shared cloud drive or a github repo so you can do the above mentioned anlaysis after i run the script.

regardless once you have said analysis, i will look through myself as well and make sure our analyses align. then i think this is safe to merge

**Comment by puririshi98** [comment]
also please update the changelog

File: examples/planetoid_train.py [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+def train_node2vec():
+    model.train()
+    total_loss = 0
+    for pos_rw, neg_rw in loader:
+        optimizer.zero_grad()
+        loss = model.loss(pos_rw.to(device), neg_rw.to(device))
+        loss.backward()
+        optimizer.step()
+        total_loss += loss.item()
+    return total_loss / len(loader)
```
Comment (akihironitta) [inline examples/planetoid_train.py:493]: Let's accumulate the loss as a tensor and not as a float so that we don't need to cuda sync every iteration:
```suggestion
    total_loss = torch.zeros(1, device=device).squeeze()
    for pos_rw, neg_rw in loader:
        optimizer.zero_grad()
        loss = model.loss(pos_rw.to(device), neg_rw.to(device))
        loss.backward()
        optimizer.step()
        total_loss += loss
    return total_loss.item() / len(loader)
```

File: examples/planetoid_train.py [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+      f'{time.perf_counter() - wall_clock_start:.4f}s')
+print('Training...')
+
+# add tensorboard
+writer = SummaryWriter()
+
+times = []
+best_val_acc = test_acc = 0.
+for epoch in range(args.epochs):
+    start = time.time()
```
Comment (akihironitta) [inline examples/planetoid_train.py:566]: nit for consistency:
```suggestion
    start = time.perf_counter()
```

**Review by akihironitta — COMMENTED** [review]
Looks great so far! ✨

Tbh, I'd vote for having many separate 90~100 lines of code examples to allow users to immediately understand what the script does and easily modify the scirpt that they're interested in (at the cost of maintaining duplicate code in the example directory), but I guess this is just my preference.

Could we also add type annotations?

File: examples/planetoid_train.py [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+    SGConv,
+    SplineConv,
+    SuperGATConv,
+    TAGConv,
+)
+from torch_geometric.typing import WITH_TORCH_SPLINE_CONV
+from torch_geometric.utils import dropout_edge
+
+if not WITH_TORCH_SPLINE_CONV:
+    quit("This example requires 'torch-spline-conv'")
```
[the remark this answers is not in the document]
Comment (xnuohz) [reply examples/planetoid_train.py]: updated and only judged when gnn_choice = splinegnn

**Comment by xnuohz** [comment]
@puririshi98 @akihironitta ready for review, thank you
1. total 14 models, 4 improved, 9 keep the same, 1 worse
[overall.csv](https://github.com/user-attachments/files/19150766/overall.csv)
3. the new file aligns config and params of the old file, when ready to merge, should delete the redundant parts, which may affect the metrics
4. the old file does not init model params explicitly, while i updated in the new file. commented out for aligning config and params of the old file
5. script for the log of new file, refer to `planetoid_train.sh`, 
[new_logs.zip](https://github.com/user-attachments/files/19150750/new_logs.zip)
6. script for the log of old file, refer to `run_cora.sh`
[old_logs.zip](https://github.com/user-attachments/files/19150751/old_logs.zip)

**Review by puririshi98 — CHANGES_REQUESTED** [review]
> Tbh, I'd vote for having many separate 90~100 lines of code examples to allow users to immediately understand what the script does and easily modify the scirpt that they're interested in (at the cost of maintaining duplicate code in the example directory), but I guess this is just my preference.



I think that is a valid point, instead of just deleting them all, @xnuohz could you leave them in and update the readme to say that

> "planetoid_train.py combines (insert list of python examples it covers) into a single unified framework that can all be run run with planetoid_train.sh. (insert description of the planetoid datasets it covers in general and explain why they can all be covered with a unified codebase)".

@akihironitta thoughts?

**Review by puririshi98 — CHANGES_REQUESTED** [review]
also @xnuohz  please figure out why that 1 example is worse and add some special condition to the framework that fixes the issue. this should be considered a strict requirement for merging

**Comment by xnuohz** [comment]
after added model params init, their results are consistent

**Comment by puririshi98** [comment]
okay lets wait till @akihironitta replies to my previous comment about restructuring this PR to satisfy his concerns.

File: CHANGELOG.md [hunk 4]
```diff
@@ -41,6 +41,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Changed
 
+- Combine 15 examples related to planetoid dataset into `planetoid_train.py` ([#10097](https://github.com/pyg-team/pytorch_geometric/pull/10097))
```
Comment (akihironitta) [inline CHANGELOG.md:44]:
```suggestion
- Combined 15 examples related to planetoid dataset into `planetoid_train.py` ([#10097](https://github.com/pyg-team/pytorch_geometric/pull/10097))
```

File: examples/planetoid_train.sh [hunk 5]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
```
Comment (akihironitta) [inline examples/planetoid_train.sh:1]: Why do we need this script?

File: torch_geometric/nn/models/pmlp.py [hunk 6]
```diff
@@ -94,7 +94,7 @@ def forward(
                 x = x.relu()
                 x = F.dropout(x, p=self.dropout, training=self.training)
 
-        return x
```
Comment (akihironitta) [inline torch_geometric/nn/models/pmlp.py:97]: Can you check whether we also apply softmax in other models? IMO, it should be consistent across models.

Also, this is a breaking change, meaning it'd require a changelog entry.

File: examples/planetoid_train.py [hunk 7]
```diff
@@ -0,0 +1,726 @@
...
+    # forward
+    if args.gnn_choice in ['splinegnn', 'gcn']:
+        out = model(data.x, data.edge_index, data.edge_attr)
+    elif args.gnn_choice == 'gcn2':
+        out = model(data.x, data.adj_t)
+    elif args.gnn_choice == 'super_gat':
+        out, att_loss = model(data.x, data.edge_index)
+    else:
+        out = model(data.x, data.edge_index)
+    # loss
```
Comment (akihironitta) [inline examples/planetoid_train.py:634]: Can you make sure the script doens't have any redundant comments like this?
```suggestion
```

**Review by akihironitta — APPROVED** [review]
I don't have a strong opinion, but as a user, I'd prefer many examples with much fewer lines rather than a single example that has ~2k lines.

I'd rather think this is a documentation issue, and so I'd suggest documenting what examples can be grouped together in a table like we have for https://github.com/pyg-team/pytorch_geometric/tree/master/examples/multi_gpu.

Feel free to disregard my comment and merge this PR if others have strong opinions :)

File: torch_geometric/nn/models/pmlp.py (hunk 6 above)
Comment (xnuohz) [reply torch_geometric/nn/models/pmlp.py:97]: all models in this example applied softmax, and also updated the changelog

File: examples/README.md [hunk 8]
```diff
@@ -22,3 +22,26 @@ For examples on scaling PyG up via multi-GPUs, see the examples under [`examples
...
+| MixHop    | 91.40%     | 75.80%     | 84.60%     |
+| SGC       | 64.00%     | 73.60%     | 77.00%     |
+| TAGCN     | 91.20%     | 73.80%     | **89.00%** |
+| GraphUNet | 90.20%     | 77.20%     | 85.20%     |
+| PMLP      | 90.40%     | **77.40%** | 84.60%     |
+| SplineGNN | 91.60%     | **77.40%** | 87.60%     |
+| GCN2      | **92.80%** | 75.60%     | 84.60%     |
+| SuperGAT  | 88.20%     | 74.40%     | 80.60%     |
+| DNA       | 90.00%     | 75.20%     | 85.80%     |
+| Node2Vec  | 84.00%     | 55.60%     | 76.80%     |
```
[the remark this answers is not in the document]
Comment (xnuohz) [reply examples/README.md]: the dataset is too small, so i removed the benchmark results from readme

**Comment by puririshi98** [comment]
@xnuohz please put all of the planetoid examples back and make the readme explain how they are all repro'd by the single planetoid_train.py if people want a unified and extendable framework. mention the originals are for individual examples and learning.

File: README.md [hunk 9]
```diff
@@ -161,15 +161,15 @@ All Graph Neural Network layers are implemented via the **[`nn.MessagePassing`](
...
+- **[ChebConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.ChebConv.html)** from Defferrard *et al.*: [Convolutional Neural Networks on Graphs with Fast Localized Spectral Filtering](https://arxiv.org/abs/1606.09375) (NIPS 2016) \[[**Example1**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/gcn.py#L36-L37), [**Example2**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/planetoid_train.py#L115-L116)\]
+- **[GATConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.GATConv.html)** from Veličković *et al.*: [Graph Attention Networks](https://arxiv.org/abs/1710.10903) (ICLR 2018) \[[**Example1**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/gat.py), [**Example2**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/planetoid_train.py)\]
```
Comment (puririshi98) [inline README.md:166]: rather than showing planetoid_train in all these duplicate places, please explain that "planetoid_train.py offers a unified codebase covering GCN, GAT, and Cheb Convs, as well as many more"

**Review by puririshi98 — CHANGES_REQUESTED** [review]
this LGTM once you address my one comment, but curious what @akihironitta thinks now. from our PoV in the longrun it is good to have unified code snippets for testing and customer use. the new planetoid_train.py offers that while explaining what old examples it combines

File: README.md [hunk 10]
```diff
@@ -183,19 +183,19 @@ These GNN layers can be stacked together to create Graph Neural Network models.
...
 - **[SignedConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.SignedConv.html)** from Derr *et al.*: [Signed Graph Convolutional Network](https://arxiv.org/abs/1808.06354) (ICDM 2018) \[[**Example**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/signed_gcn.py)\]
-- **[DNAConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.DNAConv.html)** from Fey: [Just Jump: Dynamic Neighborhood Aggregation in Graph Neural Networks](https://arxiv.org/abs/1904.04849) (ICLR-W 2019) \[[**Example**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/dna.py)\]
+- **[DNAConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.DNAConv.html)** from Fey: [Just Jump: Dynamic Neighborhood Aggregation in Graph Neural Networks](https://arxiv.org/abs/1904.04849) (ICLR-W 2019) \[[**Example1**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/dna.py)\]
```
Comment (puririshi98) [inline README.md:198]: why did you switch example -> example1 here? @xnuohz

File: README.md [hunk 11]
```diff
@@ -183,19 +183,19 @@ These GNN layers can be stacked together to create Graph Neural Network models.
...
-- **[TAGConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.TAGConv.html)** from Du *et al.*: [Topology Adaptive Graph Convolutional Networks](https://arxiv.org/abs/1710.10370) (CoRR 2017) \[[**Example**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/tagcn.py)\]
+- **[AGNNConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.AGNNConv.html)** from Thekumparampil *et al.*: [Attention-based Graph Neural Network for Semi-Supervised Learning](https://arxiv.org/abs/1803.03735) (CoRR 2017) \[[**Example1**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/agnn.py)\]
+- **[TAGConv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.TAGConv.html)** from Du *et al.*: [Topology Adaptive Graph Convolutional Networks](https://arxiv.org/abs/1710.10370) (CoRR 2017) \[[**Example1**](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/tagcn.py)\]
```
Comment (puririshi98) [inline README.md:190]: same thing here

**Review by puririshi98 — CHANGES_REQUESTED** [review]
i see you swapped example -> example1 in severl places, can you explain why?

File: README.md (hunk 10 above)
Comment (xnuohz) [reply README.md]: last time when i deleted the separate examples and updated the readme, forgot to change it back. should be fine now

**Review by puririshi98 — CHANGES_REQUESTED** [review]
LGTM, just ask that @akihironitta give a final stamp of approval

**Comment by puririshi98** [comment]
closing based on matthias's decision to not merge
